tools/kubernetes/migration.py

```diff
@@ -0,0 +1,153 @@
+    parser.add_argument("--ca-bundle", default=os.getenv("CA_BUNDLE"), help="Path to custom CA bundle file")
+    parser.add_argument('--output', '-o', default=DEFAULT_OUTPUT_DIR, help="Output directory path where pipeline YAMLs will be written(e.g., '.path/to/exported-pipelines')")
+    parser.add_argument('--namespace', default=DEFAULT_NAMESPACE, help="Namespace to filter pipelines from")
+    parser.add_argument('--batch-size', type=int, default=100,
+                    help="Number of pipelines to fetch per API call (KFP page_size). Defaults to 100.")
+    parser.add_argument('--no-pipeline-name-prefix', action='store_true',
+                        help="Disable prefixing pipeline name to version names")
+    
+    return parser.parse_args()
+
+# Fetch all pipelines from the KFP API
+def fetch_pipelines(kfp_server_host, headers, verify, namespace, batch_size):
+    url = f"{kfp_server_host}/apis/v2beta1/pipelines?namespace={namespace}&page_size={batch_size}"    
+    response = requests.get(url, headers=headers, verify=verify, timeout=DEFAULT_TIMEOUT)
+    if response.status_code == 200:
+        return response.json()
+    else:
+        raise Exception(f"Error fetching pipelines: {response.status_code} - {response.text}")
+
+# Fetch all versions for a given pipeline
+def fetch_pipeline_versions(kfp_server_host, pipeline_id, headers, verify, namespace):
+    url = f"{kfp_server_host}/apis/v2beta1/pipelines/{pipeline_id}/versions?sort_by=created_at&order_by=asc&namespace={namespace}"
+    response = requests.get(url, headers=headers, verify=verify, timeout=DEFAULT_TIMEOUT)
+    if response.status_code == 200:
+        return response.json()
+    else:
+        raise Exception(f"Error fetching versions for pipeline {pipeline_id}: {response.status_code} - {response.text}")
+
+# Creates a Kubernetes-safe name for a pipeline version by combining the pipeline name and version ID.
+# If the name is too long, it shortens it and adds index at the end to keep it unique.
+def get_version_name(pipeline_name, version_display_name, index, add_prefix):
+    if add_prefix:
+       original_version_name = f"{pipeline_name}-{version_display_name}"
+    else:
+        original_version_name = f"{version_display_name}"    
+    
+    # Clean and convert the pipeline name to be k8s-compatible
+    formatted_name = utils.maybe_rename_for_k8s(original_version_name)
+
+    if len(formatted_name) > MAX_NAME_LENGTH:
+        suffix = f"-{index}"
+        max_name_len = MAX_NAME_LENGTH - len(suffix)
+        trimmed_name = original_version_name[:max_name_len]
+        formatted_name = utils.maybe_rename_for_k8s(trimmed_name + suffix)
+
+    return formatted_name
+
+# Convert a pipeline and its versions into k8s format
+def convert_to_k8s_format(pipeline, pipeline_versions, add_prefix):
+    k8s_objects = []
+
+    # Use namespace from pipeline object if available; fallback to default
+    namespace = pipeline.get('namespace') or K8S_NAMESPACE
```

Comment:
Is the fallback for standalone mode or to detect global/sample pipelines in multiuser mode?

Reply:
@mprahl  It serves both cases, since in both scenarios the pipeline object might not have a namespace explicitly specified.